java-support/src/main/scala/io/cloudstate/javasupport/impl/controller/AnnotationBasedControllerSupport.scala

```diff
@@ -0,0 +1,462 @@
+}
+
+private class ControllerReflection(
+    val unaryHandlers: Map[String, UnaryCallInvoker],
+    val serverStreamedHandlers: Map[String, ServerStreamedCallInvoker],
+    val clientStreamedHandlers: Map[String, ClientStreamedCallInvoker],
+    val streamedHandlers: Map[String, StreamedCallInvoker]
+)
+
+private object ControllerReflection {
+  def apply(behaviorClass: Class[_], serviceMethods: Map[String, ResolvedServiceMethod[_, _]])(
+      implicit mat: Materializer
+  ): ControllerReflection = {
+
+    val allMethods = ReflectionHelper.getAllDeclaredMethods(behaviorClass)
+
+    // First, find all the call handler methods, and match them with corresponding service methods
+    val allCallHandlers = allMethods
+      .filter(_.getAnnotation(classOf[CallHandler]) != null)
+      .map { method =>
+        method.setAccessible(true)
+        val annotation = method.getAnnotation(classOf[CallHandler])
+        val name: String = if (annotation.name().isEmpty) {
+          ReflectionHelper.getCapitalizedName(method)
+        } else annotation.name()
+
+        val serviceMethod = serviceMethods.getOrElse(name, {
+          throw new RuntimeException(
+            s"Command handler method ${method.getName} for command $name found, but the service has no command by that name."
+          )
+        })
+
+        (method, serviceMethod)
+      }
+      .groupBy(_._2.name)
+      .map {
+        case (commandName, Seq((method, serviceMethod))) => (commandName, method, serviceMethod)
+        case (commandName, many) =>
+          throw new RuntimeException(
+            s"Multiple methods found for handling command of name $commandName: ${many.map(_._1.getName).mkString(", ")}"
+          )
+      }
+
+    val unaryCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if !serviceMethod.descriptor.isClientStreaming && !serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new UnaryCallInvoker(method, serviceMethod)
+    }.toMap
+
+    val serverStreamedCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if !serviceMethod.descriptor.isClientStreaming && serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new ServerStreamedCallInvoker(method, serviceMethod)
+    }.toMap
+
+    val clientStreamedCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if serviceMethod.descriptor.isClientStreaming && !serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new ClientStreamedCallInvoker(method, serviceMethod, mat)
+    }.toMap
+
+    val streamedCallHandlers = allCallHandlers.collect {
+      case (commandName, method, serviceMethod)
+          if serviceMethod.descriptor.isClientStreaming && serviceMethod.descriptor.isServerStreaming =>
+        commandName -> new StreamedCallInvoker(method, serviceMethod, mat)
+    }.toMap
+
+    ReflectionHelper.validateNoBadMethods(
```

Comment:
just a small change. perhaps it can be done first to fail fast after allMethods are processed.